src/shell-interface.h

```diff
@@ -87,17 +87,23 @@ struct ShellExternalInterface : ModuleInstance::ExternalInterface {
   } memory;
 
   std::unordered_map<Name, std::vector<Literal>> tables;
+  std::map<Name, ModuleInstance*> linkedInstances;
 
   ShellExternalInterface() : memory() {}
   virtual ~ShellExternalInterface() = default;
 
-  void init(Module& wasm, ModuleInstance& instance) override {
-    memory.resize(wasm.memory.initial * wasm::Memory::kPageSize);
-
+  void init(Module& wasm,
+            ModuleInstance& instance,
+            std::map<Name, std::shared_ptr<ModuleInstance>> linkedInstances =
+              {}) override {
+    linkedInstances.swap(linkedInstances);
+    if (wasm.memory.exists && (!wasm.memory.imported() || wasm.memory.module == "env")) {
```

Comment:
This looks a bit arbitrary. I needed this for `ctor-eval` tests, where a memory may be imported from `env`. Another approach would've been to instantiate an artificial `env` module and link to those test modules, but that was even more complicated.